### PR TITLE
[Feature] - Add functionality to overlay high res center onto output images

### PR DESCRIPTION
**Pull request by zekeflaton**
Adds a new function that allows for zooming in on the center of the image with some variability to get a higher resolution center.  This will then paste that output into the final image.

### CODE REVIEW

File: semantic/unet/utils.py [hunk 1]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
     labels_img.putalpha(floor(255 * alpha))
     print(img.size, labels_img.size)
     return Image.alpha_composite(img, labels_img)
+
+
+def get_image_with_high_res_center(model, img : Image, zoom : float = 2, output_dir: str = "output") -> np.ndarray:
+    # Create the output directory if it does not exist
+    if not os.path.exists(output_dir):
```
Comment (hlfshell) [inline semantic/unet/utils.py]: This shouldn't be needed. It should return a PIL, not an array, and it shouldn't write to disk at all.

File: semantic/unet/utils.py [hunk 2]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+        os.makedirs(output_dir)
+
+    # Get input size of network
+    input_size = model.layers[0].get_output_at(0).get_shape().as_list()
```
Comment (hlfshell) [inline semantic/unet/utils.py]: make use of `infer` so you don't have to repeat this step throughout with tracking `input_size`

File: semantic/unet/utils.py [hunk 3]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+    final_image.save(os.path.join(output_dir, "final_image1.png"))  # Output of original image
+    final_image_cropped.save(os.path.join(output_dir, "final_image_2.png"))  # Output of cropped image
+
+    # Paste the cropped image into the center of the original image and save it
+    img_w, img_h = final_image_cropped.size
+    bg_w, bg_h = final_image.size
+    offset = ((bg_w - img_w) // 2, (bg_h - img_h) // 2)
+    fi3 = final_image
+    fi3.paste(final_image_cropped, offset)
+    fi3.save(os.path.join(output_dir, "fi3.png"))
```
Comment (hlfshell) [inline semantic/unet/utils.py]: This function shouldn't save anything. Same goes for all save's above.

File: semantic/unet/utils.py [hunk 4]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
     labels_img.putalpha(floor(255 * alpha))
     print(img.size, labels_img.size)
     return Image.alpha_composite(img, labels_img)
+
+
+def get_image_with_high_res_center(model, img : Image, zoom : float = 2, output_dir: str = "output") -> np.ndarray:
```
Comment (hlfshell) [inline semantic/unet/utils.py]: Is this even needed with the other function?

File: semantic/unet/utils.py [hunk 5]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+    # print(img.size)
+
+    # Determine the crop points based on the zoom level
+    crop_points = (
+        int(height/2-height/(2*zoom)),
+        int(width/2-width/(2*zoom)),
+        int(height/2+height/(2*zoom)),
+        int(width/2+width/(2*zoom))
+    )
+    # print(crop_points)
```
Comment (hlfshell) [inline semantic/unet/utils.py]: Unused commented code.

File: semantic/unet/utils.py [hunk 6]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+    # [batch_size (None), width, height, channels]. Really we just want
+    # width and height:
+    input_size = (input_size[1], input_size[2])
+
+    # Get the low res image as an input for the model
+    low_res_input = rgb_image_to_input(img, input_size=input_size)
+
+    # Get the dimensions of the original image
+    height, width = img.size
+    # print(img.size)
```
Comment (hlfshell) [inline semantic/unet/utils.py]: Unused commented code.

File: semantic/unet/utils.py [hunk 7]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+        # 0 is white, 255 is black.  White pixels will allow for overwrites
+        mask = np.argmax(overlay_labels, axis=-1)
+        mask = np.isin(mask, [1, 6, 7]).astype(int)*255  # only update pixels where the label value is specified
+        mask = resize(mask, overlay_labels_image.size[::-1], order=0, preserve_range=True, anti_aliasing=False)
+
+        # Convert numpy mask to pillow image
+        mask_img = Image.fromarray(mask)
+        mask_img = mask_img.convert("1")  # Convert to 1 channel, needed for paste below
+
+        if output_dir:
```
Comment (hlfshell) [inline semantic/unet/utils.py:136]: No writing in this function, no saving. It does its one thing and then returns it to another function.

File: semantic/unet/utils.py [hunk 8]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+            # Save component images
+            mask_img.save(os.path.join(output_dir, "mask_{}.png".format(i)))
```
Comment (hlfshell) [inline semantic/unet/utils.py:138]: No saving here.

File: semantic/unet/utils.py [hunk 9]
```diff
@@ -73,3 +74,129 @@ def overlay_labels_on_input(img: Image, labels : np.ndarray, alpha : float = 0.4
...
+
+    # Paste the cropped image into the center of the original image and save it
+    img_w, img_h = final_image_cropped.size
+    bg_w, bg_h = final_image.size
+    offset = ((bg_w - img_w) // 2, (bg_h - img_h) // 2)
+    fi3 = final_image
+    fi3.paste(final_image_cropped, offset)
+    fi3.save(os.path.join(output_dir, "fi3.png"))
+
+def stitch_high_res_overlays_together(model, img : Image, sub_sections : int = 4, output_dir: str = None) -> np.ndarray:
```
Comment (hlfshell) [inline semantic/unet/utils.py]: Since this returns a PIL image, the type hint should say as such above

**Review by hlfshell — CHANGES_REQUESTED** [review]

File: semantic/unet/utils.py (hunk 9 above)
Comment (zekeflaton) [reply semantic/unet/utils.py]: right, done.

File: semantic/unet/utils.py (hunk 3 above)
Comment (zekeflaton) [reply semantic/unet/utils.py]: yup, forgot about this one.  deleting it in favor of the new sliding window function.

File: semantic/unet/utils.py (hunk 8 above)
Comment (zekeflaton) [reply semantic/unet/utils.py:138]: Seems harmless to have this behind an argument.  Lets use save images if we want to use it for debugging/one-offs or skip it entirely to use in a more automated fashion.

**Review by hlfshell — APPROVED** [review]

**Pull request merged**